data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,545 @@
+}
+
+impl AsRef<str> for BoxedString {
+    #[inline]
+    fn as_ref(&self) -> &str {
+        self.0.as_ref()
+    }
+}
+
+impl From<&str> for BoxedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn from(value: &str) -> Self {
+        let payload = value.as_bytes().to_vec().into_boxed_slice();
+        // SAFETY: correct string.
+        Self(unsafe { from_boxed_utf8_unchecked(payload) })
+    }
+}
+
+impl From<String> for BoxedString {
+    #[inline]
+    fn from(value: String) -> Self {
+        Self(value.into_boxed_str())
+    }
+}
+
+#[derive(Clone, Copy)]
+#[repr(C)]
+struct InlinedString {
+    #[cfg(target_endian = "big")]
+    payload: [u8; MAX_INLINED_STRING_LEN],
+    /// Least-significant bit is always 1 to distinguish inlined variant.  
+    len: u8,
+    #[cfg(target_endian = "little")]
+    payload: [u8; MAX_INLINED_STRING_LEN],
+}
+
+impl InlinedString {
+    #[inline]
+    const fn len(self) -> usize {
+        (self.len >> 1_u8) as usize
+    }
+
+    #[inline]
+    const fn is_inlined(self) -> bool {
+        self.len % 2 > 0
+    }
+
+    #[inline]
+    const fn new() -> Self {
+        Self {
+            payload: [0; MAX_INLINED_STRING_LEN],
+            // Set least-significant bit to mark inlined variant.
+            len: 1,
+        }
+    }
+}
+
+impl AsRef<str> for InlinedString {
+    #[allow(unsafe_code)]
+    #[inline]
+    fn as_ref(&self) -> &str {
+        // SAFETY: created from valid utf-8.
+        unsafe { core::str::from_utf8_unchecked(&self.payload[..self.len()]) }
+    }
+}
+
+impl<'value> TryFrom<&'value str> for InlinedString {
+    type Error = &'value str;
+
+    #[allow(clippy::cast_possible_truncation)]
+    #[inline]
+    fn try_from(value: &'value str) -> Result<Self, Self::Error> {
+        let len = value.len();
+        if len > MAX_INLINED_STRING_LEN {
+            return Err(value);
+        }
+        let mut inlined = Self::new();
+        inlined.payload.as_mut()[..len].copy_from_slice(value.as_bytes());
+        // Truncation won't happen because we checked that the length shorter than `MAX_INLINED_STRING_LEN`.
+        // Set least-significant bit to mark inlined variant.
+        inlined.len += (len << 1_usize) as u8;
+        Ok(inlined)
+    }
+}
+
+impl TryFrom<String> for InlinedString {
+    type Error = String;
+
+    #[inline]
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        match Self::try_from(value.as_str()) {
+            Ok(inlined) => Ok(inlined),
+            Err(_) => Err(value),
+        }
+    }
+}
+
+impl Debug for InlinedString {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        f.debug_struct("InlinedString")
+            .field("len", &self.len())
+            .field("payload", &self.as_ref())
+            .finish()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use core::mem::{align_of, size_of};
+    use std::collections::hash_map::DefaultHasher;
+
+    use super::*;
+
+    fn run_with_strings(f: impl Fn(String)) {
```

Comment:
Like this function, since you can now add more strings to the iterator. I would suggest adding strings that contain UTF-8 characters here and seeing if the tests keep working. Consider the cases of 4-byte characters. many 2-bytes, ASCII all the way except last, odd number of UTF-8, wonky alignment…